packages/ember-toucan-form/src/-private/file-input-field.ts

```diff
@@ -0,0 +1,64 @@
+import Component from '@glimmer/component';
+import { assert } from '@ember/debug';
+import { action } from '@ember/object';
+
+import type { HeadlessFormBlock, UserData } from './types';
+import type { ToucanFormFileInputFieldComponentSignature as BaseFileInputFieldSignature } from '@crowdstrike/ember-toucan-core/components/form/fields/file-input';
+import type { FormData, FormKey, ValidationError } from 'ember-headless-form';
+
+export interface ToucanFormFieldInputFieldComponentSignature<
+  DATA extends UserData,
+  KEY extends FormKey<FormData<DATA>> = FormKey<FormData<DATA>>
+> {
+  Element: HTMLInputElement;
+  Args: Omit<
+    BaseFileInputFieldSignature['Args'],
+    'error' | 'files' | 'onChange' | 'name'
+  > & {
+    /**
+     * The name of your field, which must match a property of the `@data` passed to the form
+     */
+    name: KEY;
+
+    /*
+     * @internal
+     */
+    form: HeadlessFormBlock<DATA>;
+  };
+  Blocks: BaseFileInputFieldSignature['Blocks'];
+}
+
+export default class ToucanFormFileInputFieldComponent<
+  DATA extends UserData,
+  KEY extends FormKey<FormData<DATA>> = FormKey<FormData<DATA>>
+> extends Component<ToucanFormFieldInputFieldComponentSignature<DATA, KEY>> {
+  hasOnlyLabelBlock = (hasLabel: boolean, hasHint: boolean) =>
+    hasLabel && !hasHint;
+  hasHintAndLabelBlocks = (hasLabel: boolean, hasHint: boolean) =>
+    hasLabel && hasHint;
+  hasLabelArgAndHintBlock = (hasLabel: string | undefined, hasHint: boolean) =>
+    hasLabel && hasHint;
+
+  mapErrors = (errors?: ValidationError[]) => {
+    if (!errors) {
+      return;
+    }
+
+    // @todo we need to figure out what to do when message is undefined
+    return errors.map((error) => error.message ?? error.type);
+  };
+
+  @action
+  assertArrayOfFiles(value: unknown): File[] | undefined {
+    assert(
+      `Only File[] values are expected for ${String(
+        this.args.name
+      )}, but you passed ${typeof value}`,
+      typeof value === 'undefined' ||
+        Array.isArray(value) ||
+        (value as File[]).forEach((file) => file instanceof File)
```

Comment:
omg, excellent catch 🙈.  One thing that was adjusted with file-input is instead of returning a `FileList`, it returns an Array of files explicitly due to some limitations.  I can't remember quite why off the top of my head, but maybe @nicolechung remembers and can chime in.

---

For some code golf, maybe:

```
typeof value === 'undefined' ||
        Array.isArray(value) ||
        (value as File[]).map((file) => file instanceof File).every((file) => Boolean(file))
```

---

Or maybe -- do we only care if they provide us an array or undefined and we don't need to get this granular with the types?

```
typeof value === 'undefined' || Array.isArray(value)
```

Reply:
I'd be curious to hear what @nicolechung says. The custom type (`File[]`) in lieu of its native representation is unfortunate.

Either way, if you go with what you have above, you could golf `typeof value === 'undefined'` down to `value === undefined`. And do you need to `map` and the cast? Or could you reduce it `(value).every((file) => file instanceof File))`?

Reply:
Doh, I'm dumb. You're right the single every will do!